### ISSUE TITLE
Why are the keys strings?

### ISSUE THREAD

**Issue by tedkornish** [patrickmn/go-cache#26]
Hey there,

A coworker and I were wondering about the reasoning behind making keys to the cache strings. Using structs as keys in maps is slower and often less convenient than using structs (as Ashish Gandhi details towards the end of [this talk](https://www.hakkalabs.co/articles/optimizing-go-3k-requestssec-480k-requestssec)). Is it because allowing for a cache key of `interface{}` introduces the risk of runtime type error if a user tried to store, say, a slice?

Thanks for writing and maintaining the package!

T

**Comment by patrickmn** [comment]
I wrote a couple benchmarks to see the performance difference, and a map of interface{} is over twice as slow on my Lenovo X250 i7-5600U 2.60GHz (compared to a map of strings):

BenchmarkRWMutexMapGet-4                	50000000	        25.6 ns/op
BenchmarkRWMutexInterfaceMapGetStruct-4 	30000000	        59.8 ns/op
BenchmarkRWMutexInterfaceMapGetString-4 	20000000	        72.8 ns/op

which makes sense when you consider that interface{} values require what is essentially a pointer indirection at runtime.

Further, although interface{} values as keys might be convenient in a few cases, they severely reduce the readability of the documentation. I imagine the memory overhead would be higher as well.

What I can gather from briefly looking at the talk is that doing string manipulation is expensive because it causes allocations. It's not something inherently about using strings as keys, but the performance penalty of constructing those keys since strings are immutable. I think if you are at a point where this is a problem, go-cache isn't actually right for you: it would be much faster to have each goroutine manage its own map without locks/synchronization and using a map[specific type]value. That being said, using a []byte to construct the key, then converting it to a string is pretty efficient.

In short, I'm not seeing a convincing argument for interface{} keys since 1. they're a fairly niche use case, and 2. I can't think of a scenario where you wouldn't rather want a map to a specific type, which in Go unfortunately means that you have to implement the cache yourself. Are there specific examples where you think this would be useful?

FWIW, the LRU cache from groupcache uses interface{} for its keys: https://github.com/golang/groupcache/blob/master/lru/lru.go

**Comment by Freeaqingme** [comment]
Did you try to use interface{} or *interface{} ? There should not be any runtime overhead as long as you're not using type assertions.

**Comment by patrickmn** [comment]
Unless something's changed, any interface value is essentially a struct with a pointer to information about the type or its method table, and a pointer to the actual data, e.g. string. From http://research.swtch.com/interfaces: "Interface values are represented as a two-word pair giving a pointer to information about the type stored in the interface and a pointer to the associated data. Assigning b to an interface value of type Stringer sets both words of the interface value."

i.e. interface values are boxed. An *interface{} value would be doubly boxed. Both are slower than using strings without indirection.

You can run the benchmarks yourself with
`go test github.com/patrickmn/go-cache --bench=BenchmarkRW`

You're looking for BenchmarkRWMutexMapGet, BenchmarkRWMutexInterfaceMapGetStruct (map[interface{}]interface{} that uses struct keys), and BenchmarkRWMutexInterfaceMapGetString (map[interface{}]interface{} that uses string keys): https://github.com/patrickmn/go-cache/blob/master/cache_test.go#L1460-L1487

The latter two tests are much slower than the first (which uses a map[string]interface{})

**Comment by patrickmn** [comment]
So I think the gist here is:

  - You would need a map[specific struct type]value to avoid allocations (and indirection,) which means implementing the eviction and synchronization for a custom map yourself. A library can't do this (at least in Go 1.)

  - String building can be expensive, but string([]byte) addresses this pretty well.

I get that struct keys (and int keys, for that matter) can be useful in specific cases, but I don't think interface{} keys are the solution. It is certainly not worth degrading performance in the vast majority of use cases (which use strings) over.

**Comment by wangyoucao577** [comment]
I need a thread-safe(goroutine-safe) map with `int64` as key. It seems that the best solution might be that implement my own cache by tie `map[int64]value` and `RWMutex` together. But I might need another thread-safe map with `int64,int64` but same logic later. Should I implement another one too?  

I saw several possible solutions, but all of them have drawbacks: 
- implement both of them: duplicate logic implementation
- use this `go-cache`: have to convert `int64` or `int64,int64` to `string`, it needs allocation for every edit/query 
- use https://golang.org/pkg/sync/#RWMutex: it uses `interface{}` as key, from the discussion above, it'll be slower too due to type assertion. 

Any suggestion? Thanks!
